spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -359,31 +359,37 @@ public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
 		}
 
 		private boolean subscribeLock(long time) throws ExecutionException, InterruptedException {
-			if (!obtainLock()) {
-				if (!RedisLockRegistry.this.redisMessageListenerContainer.isRunning()) {
-					RedisLockRegistry.this.redisMessageListenerContainer.afterPropertiesSet();
-					RedisLockRegistry.this.redisMessageListenerContainer.start();
-				}
-				try {
-					Future<String> future =
-							RedisLockRegistry.this.unlockNotifyMessageListener.subscribeLock(this.lockKey);
-					//DCL
-					if (!obtainLock()) {
-						try {
-							//if short expireAfter key expire for ttl, no receive unlock msg
-							long waitTime = time >= 0 ? time : RedisLockRegistry.this.expireAfter;
-							future.get(waitTime, TimeUnit.MILLISECONDS);
-						}
-						catch (TimeoutException ignore) {
+			final long expiredTime = System.currentTimeMillis() + time;
+			while (time == -1 || expiredTime >= System.currentTimeMillis()) {
```

Comment:
Wait a minute, This code is changing.
It seems better to move while to the bottom.